cpp/include/raft/cudart_utils.h

```diff
@@ -0,0 +1,211 @@
+    oss << std::endl << "Obtained " << depth << " stack frames" << std::endl;
+    char** strings = backtrace_symbols(stack, depth);
+    if (strings == nullptr) {
+      oss << "But no stack trace could be found!" << std::endl;
+      msg_ += oss.str();
+      return;
+    }
+    ///@todo: support for demangling of C++ symbol names
+    for (int i = 0; i < depth; ++i) {
+      oss << "#" << i << " in " << strings[i] << std::endl;
+    }
+    free(strings);
+    msg_ += oss.str();
+#endif  // __GNUC__
+  }
+};
+
+/** macro to throw a runtime error */
+#define THROW(fmt, ...)                                                        \
+  do {                                                                         \
+    std::string msg;                                                           \
+    char errMsg[2048]; /* NOLINT */                                            \
+    std::snprintf(errMsg, sizeof(errMsg),                                      \
+                  "exception occured! file=%s line=%d: ", __FILE__, __LINE__); \
+    msg += errMsg;                                                             \
+    std::snprintf(errMsg, sizeof(errMsg), fmt, ##__VA_ARGS__);                 \
+    msg += errMsg;                                                             \
+    throw raft::exception(msg);                                                \
+  } while (0)
+
+/** macro to check for a conditional and assert on failure */
+#define ASSERT(check, fmt, ...)              \
+  do {                                       \
+    if (!(check)) THROW(fmt, ##__VA_ARGS__); \
+  } while (0)
+
+/** check for cuda runtime API errors and assert accordingly */
+#define CUDA_CHECK(call)                                               \
+  do {                                                                 \
+    cudaError_t status = call;                                         \
+    ASSERT(status == cudaSuccess, "FAIL: call='%s'. Reason:%s", #call, \
+           cudaGetErrorString(status));                                \
+  } while (0)
+
+///@todo: enable this only after we have added logging support in raft
+// /**
+//  * @brief check for cuda runtime API errors but log error instead of raising
+//  *        exception.
+//  */
+// #define CUDA_CHECK_NO_THROW(call)                                            \
+//   do {                                                                       \
+//     cudaError_t status = call;                                               \
+//     if (status != cudaSuccess) {                                             \
+//       RAFT_LOG_ERROR("CUDA call='%s' at file=%s line=%d failed with %s ",    \
+//                      #call, __FILE__, __LINE__, cudaGetErrorString(status)); \
+//     }                                                                        \
+//   } while (0)
+
+/** helper method to get max usable shared mem per block parameter */
+inline int get_shared_memory_per_block() {
+  int devId;
```

Comment:
We're mostly using lower_case for variable names (except for const ones), but here we're using CamelCase. Are we allowing to mix these two styles, or is there another rule I am missing?

Reply:
It seems functions are lower_case and internal variables tend to be CamelCase

Reply:
that's my bad. I missed setting up the `LocalVariable*` options for the `readability-identifier-naming` check! Thanks for catching this. Will update soon with the fixes.

Reply:
This should be done now.